enhancements/hypershift/node-lifecycle.md

```diff
@@ -0,0 +1,198 @@
+
+It runs a "token" Secret controller underneath.
+This is the controller that generates and caches the Ignition payload for any pair of `.release` and `.config` input for each NodePool.
+The TokenSecret controller watches "token" Secrets created by the NodePool controller and:
+- Generates the Ignition payload for that `.release` and `.config` pair (See below for details).
+- Stores the payload in memory indexed by "token" and exposes it to the Server.
+- Sets a TTL for the token.
+- After TTL/2 move the original token to rotated spot and create a new one. This results in both tokens for the same `.release` and `.config` pair coexisting during a TTL/2 duration. This give in flight operations e.g. Host booting with an userdata pointing to the rotated token.
+- Removes the token from memory after the TTL.
+
+Payload generation:
+- When generating a new payload is needed the TokenSecret controller fetches the MCO binaries for the given `.release` image.
+- It runs mco controllers in bootstrap mode.
+- It runs mcs in bootstrap mode and makes a local http query to get the payload served.
+- It stores the payload by "token" for the Ignition server (So it can be consumed by new Machines i.e. scale up or MachineDeployment rolling upgrade operations).
+- It stores the payload in a key of the token Secret (So it can be used by the Inplace upgrader i.e. upgrade of existing Machines belonging to a MachineSet).
+
+Deletion:
+- When a NodePool `.release` or `.config` is changed, a new token secret is created by the NodePool controller that corresponds to the new pair.
+- The old token Secret is marked with an expiration date by the NodePool controller.
+- Once the expiration date has passed, the old token Secret is deleted by the TokenSecret controller and its counterpart tokens in the Ignition Server memory are deleted. This strategy is done to allow in flight provisions that occurred in proximity to the nodePool upgrade to complete.
+
+#### Inplace Upgrader
+The Inplace Upgrader is a controller that runs in the control plane namespace and watches both MachineSets (management cluster) and Nodes (guest cluster).
+It orchestrates in place upgrades for a pool of Nodes i.e. MachineSets honouring the following design invariants:
+- Draining and any other step of the process but the upgrade itself are performed in a centralised fashion. This reduces the risk for high privilege Service Accounts permissions to be exposed in Nodes and therefore being vulnerable to containers escape.
+- Prevent the cluster end user as much as possible from signaling back interfering in the product owned behaviour.
+
+During the in place upgrade process as of today annotations are used to signal back and forward between NodePool<->MachineSets and between MachineSets<->Nodes.
+This might in future evolve to a different form of contract e.g. CRD.
+
+Workflow:
+- The NodePool controller signal an upgrade (`.release` and `.config` pair change) annotation the underlying MachineSet.
```

Comment:
s/controller signal/controller signals/
s/annotation the underlying MachineSet/via an annotation in the underlying MachineSet/